### PR TITLE
IGNITE-955 Local listener in continuous queries should not be mandatory

### PR DESCRIPTION
**Pull request by voipp**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java [hunk 1]
```diff
@@ -532,8 +532,11 @@ private ClusterGroup projection(boolean loc) {
         if (qry instanceof ContinuousQuery) {
             ContinuousQuery<K, V> qry0 = (ContinuousQuery<K, V>)qry;
 
-            if (qry0.getLocalListener() == null)
-                throw new IgniteException("Mandatory local listener is not set for the query: " + qry);
+            if (qry0.getLocalListener() == null &&
```
[the remark this answers is not in the document]
Comment (voipp) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java]: in ticket we should make local listener optional if only remote filter is set to return false

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java [hunk 2]
```diff
@@ -545,8 +548,10 @@ private ClusterGroup projection(boolean loc) {
         else {
             ContinuousQueryWithTransformer<K, V, ?> qry0 = (ContinuousQueryWithTransformer<K, V, ?>)qry;
 
-            if (qry0.getLocalListener() == null)
-                throw new IgniteException("Mandatory local transformed event listener is not set for the query: " + qry);
+            if (qry0.getLocalListener() == null && qry0.getRemoteFilterFactory() == null) {
```
[the remark this answers is not in the document]
Comment (voipp) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java]: in ticket we should make local listener optional if only remote filter is set to return false

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryHandler.java [hunk 3]
```diff
@@ -933,8 +933,6 @@ private void notifyLocalListener(Collection<CacheEntryEvent<? extends K, ? exten
         @Nullable IgniteClosure<CacheEntryEvent<? extends K, ? extends V>, ?> trans) {
         EventListener locTransLsnr = localTransformedEventListener();
 
-        assert (locLsnr != null && locTransLsnr == null) || (locLsnr == null && locTransLsnr != null);
```
[the remark this answers is not in the document]
Comment (voipp) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryHandler.java]: yeah, thx

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 4]
```diff
@@ -1138,6 +1144,94 @@ public void testExpired() throws Exception {
         }
     }
 
+    /**
```
[the remark this answers is not in the document]
Comment (voipp) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: yeah

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 5]
```diff
@@ -1138,6 +1144,94 @@ public void testExpired() throws Exception {
...
+     *
+     */
+    public void testContinuousQueryRemoteFilterFactoryWithoutLocalLstnr() {
```
[the remark this answers is not in the document]
Comment (voipp) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (voipp) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: nope(

**Comment by dspavlov** [comment]
The ticket is resolved, closing PR. Feel free to reopen if needed.

**Pull request closed without merging**